### DISCUSSION TITLE
Special/Unicode Characters Not Parsing Properly

### DISCUSSION THREAD

**Discussion by sagargg** [dathere/datapusher-plus#110] [Q&A] [answered]
I have encountered an issue where special or Unicode characters within the values of certain data eg.  €  are not parsing properly.  

Example of CSV:
```
Name,Age,City,Amount
John,25,New York,$2000
Jane,30,London, €2000
```

On Data Preview:
<img width="535" alt="Screenshot 2023-06-27 at 10 19 10 AM" src="https://github.com/dathere/datapusher-plus/assets/87696933/68bd32fb-ca7d-47b0-9622-ce5e6391c795">

On Data API:
<img width="311" alt="Screenshot 2023-06-27 at 10 20 00 AM" src="https://github.com/dathere/datapusher-plus/assets/87696933/b19684ee-0257-4b3b-8189-7ee96e7686c5">

On qsvdp CLI: 

<img width="795" alt="Screenshot 2023-06-27 at 10 20 59 AM" src="https://github.com/dathere/datapusher-plus/assets/87696933/b129dd51-48f0-4a74-a6f5-5f5c0c03c0ef">

**Comment by TomeCirun** [comment]
<img width="549" alt="image" src="https://github.com/dathere/datapusher-plus/assets/72216462/b2ae5bdc-3558-470e-888a-64e75314ecef">

![image](https://github.com/dathere/datapusher-plus/assets/72216462/b7a4de67-bcd4-4534-917e-9dd0333a3be2)

Working fine on the `dev-v1.0` branch with `QSV` version `0.91.0` @sagargg 

We are currently testing and preparing the DP+ for a version 1.0 release
more info [here](https://github.com/dathere/datapusher-plus/issues/98)

**Comment by jqnatividad** [comment]
Hi @sagargg ,
Can you confirm what version of DP+ you're using? 0.15.0 should also work without issues.

If its still a problem, what's the encoding of the file you're uploading?

DP+ normalizes non UTF8-encoded files to their "lossy" encoding - i.e. replacing invalid UTF-8 sequences with [U+FFFD REPLACEMENT CHARACTER](https://doc.rust-lang.org/std/char/constant.REPLACEMENT_CHARACTER.html), which is the � character you're getting instead of the euro currency symbol.

**Comment by sagargg** [comment]
@jqnatividad  Maybe it is happening only on macOS. 
<img width="1200" alt="Screenshot 2023-07-07 at 9 30 47 AM" src="https://github.com/dathere/datapusher-plus/assets/87696933/ac57ab40-59fd-4658-acf5-8ac1c7ac90ed">


Even it is happening on the latest version of qsv.

**Comment by jqnatividad** [comment]
@sagargg , do you mind uploading a sample file here?

I also have macOS (M2 Pro Mac Mini) and it works without issues.

[testeuro.csv](https://github.com/dathere/datapusher-plus/files/11974440/testeuro.csv)


```
$ qsvdp --version
qsvdp 0.108.0-standard-Luau 0.581;self_update-12-12;25.60 GiB-0B-0B-32.00 GiB (aarch64-apple-darwin compiled with Rust 1.70.0) prebuilt

$ qsvdp input testeuro.csv 
Name,Age,City,Amount
John,25,New York,$2000
Jane,30,London,€2000

$ qsvdp sniff testeuro.csv 
Path: /Users/joelnatividad/Local/GitHub/qsv/testeuro.csv
Sniff Timestamp: 2023-07-07T04:07:46.429694+00:00
Last Modified: 2023-07-07T04:06:33+00:00
Delimiter: ,
Header Row: true
Preamble Rows: 0
Quote Char: none
Flexible: false
Is UTF8: true
Detected Mime Type: application/csv
Retrieved Size (bytes): 67
File Size (bytes): 67
Sampled Records: 2
Estimated: false
Num Records: 2
Avg Record Len (bytes): 19
Num Fields: 4
Stats Types: false
Fields:
    0:  Text      Name
    1:  Unsigned  Age
    2:  Text      City
    3:  Text      Amount
```

**Comment by jqnatividad** [comment]
BTW, I just noticed you're on a very old version of qsv - 0.92.0.

You should upgrade to qsv 0.108.0

**Comment by sagargg** [comment]
@jqnatividad   yeah, it worked with the file that you've shared.  
<img width="701" alt="Screenshot 2023-07-07 at 10 00 42 AM" src="https://github.com/dathere/datapusher-plus/assets/87696933/b978c50a-e1f7-40cb-b1aa-2fb3e8ee9d72">


Here is the file i'm using.
[csv.csv](https://github.com/dathere/datapusher-plus/files/11974449/csv.csv)

**Comment by jqnatividad** [comment, marked as answer]
The encoding of the file is not utf-8, that's why `qsv input` did a lossy-transcoding to utf8.

```
$ file -I csv.csv
csv.csv: text/plain; charset=iso-8859-1

$ file -I testeuro.csv 
testeuro.csv: text/csv; charset=utf-8
```

**Comment by sagargg** [comment]
okay. Now I understood, Thank you for clarifying.
